jobserver/api/jobs.py

```diff
@@ -234,6 +236,73 @@ class Meta:
         model = Workspace
 
 
+class JobRequestAPIListCreate(APIView):
+    """
+    Dispatcher view for the /job-requests/ endpoint.
+
+    The handlers for the GET and POST verbs are very different in this
+    endpoint.  Those differences are encapsulated in the Views we hand off to
+    in this one.
+    """
+
+    def dispatch(self, request, *args, **kwargs):
+        if request.method.lower() == "post":
+            return JobRequestAPICreate().dispatch(request, *args, **kwargs)
+
+        return JobRequestAPIList().dispatch(request, *args, **kwargs)
+
+
+class JobRequestAPICreate(APIView):
+    authentication_classes = []
+
+    class serializer_class(serializers.ModelSerializer):
+        backend = serializers.SlugRelatedField(
+            slug_field="slug", queryset=Backend.objects.all()
+        )
+        workspace = serializers.SlugRelatedField(
+            slug_field="name", queryset=Workspace.objects.all()
+        )
+
+        class Meta:
+            fields = [
+                "backend",
+                "workspace",
+                "sha",
+                "project_definition",
+                "requested_actions",
+                "force_run_dependencies",
+            ]
+            model = JobRequest
+
+        def validate_project_definition(self, value):
+            load_pipeline(value)
+
+            return value
+
+    def post(self, request, *args, **kwargs):
+        auth_header = request.headers.get("Authorization")
```

Comment:
How come this is checking the token in a different way to the APIViews below? Only asking because I'm trying to learn how the authentication works and the views below use an initial method..

Reply:
We have 4 ways of authenticating with our API:
* no authentication (for the public list of JobRequests)
* backend token (mostly for a job-runner instance to sync with job-server)
* session auth (for the output viewer)
* bot user token (for OSI)

Typically in an APIView auth[n|z] is checked in the `initial` method (which most of our views do), because it fires before the relevant verb handler method.  But in this case we should never get a non-POST request because of how we invoke the view and I thought it was simpler to reduce the view to just the POST handler method.

Reply:
Ah ha, thank you. That makes sense. I think I was also trying to understand the difference between the backend token and the bot user token, which I found confusing on the first read.